middle_end/flambda2/types/env/aliases.ml

```diff
@@ -982,89 +982,96 @@ let add ~binding_time_resolver ~binding_times_and_modes t
       ~original_t add_result;
   add_result
 
-(* CR mshinwell: This needs documenting. For the moment we allow relations
-   between canonical elements that are actually incomparable under the name mode
-   ordering, and check in [get_canonical_element_exn] accordingly. However maybe
-   we should never allow these situations to arise. *)
-(* let canonical_mode = name_mode t add_result.canonical_element in let
-   alias_of_mode = name_mode t add_result.alias_of in match
-   Name_mode.compare_partial_order canonical_mode alias_of_mode with | Some _ ->
-   add_result | None -> Misc.fatal_errorf "Canonical %a has mode incomparable
-   with %a in:@ %a" Simple.print add_result.canonical_element Simple.print
-   add_result.alias_of print t *)
+(* CR mshinwell: For the moment we allow relations between canonical elements
+   that are actually incomparable under the name mode ordering, and check in
+   [get_canonical_element_exn] accordingly. However maybe we should never allow
+   these situations to arise. *)
+
+let find_earliest_alias t ~canonical_element ~binding_times_and_modes
+    ~min_binding_time ~min_name_mode ~binding_time_resolver
+    ~coercion_from_canonical_to_element =
+  (* There used to be a shortcut that avoided consulting the aliases in the
+     common case that [element] is itself canonical and has no aliases, since
+     then it does not appear in [canonical_elements]. However, this shortcut was
+     broken: a canonical element *with* known aliases may still not appear in
+     [canonical_elements]. See tests/flambda2-aliases for a test that gave
```

Comment:
(This is not related to this pull request, but I cannot find `tests/flambda2-aliases`.)

Reply:
(Maybe it is https://github.com/ocaml-flambda/ocaml/blob/flambda2.0-stable/testsuite/tests/flambda2-aliases/test.ml.)

Reply:
Thanks, I've added a CR for this test to be imported later.